### PR TITLE
search: add stable parameter

### PR DESCRIPTION
**Pull request by rvantonder**
Stacked on #9680. Adds `stable` parameter to ensure deterministic result ordering. Uses pagination underneath the hood.

**Notes**

---

@slimsag This will use a default result count of 30, since that's in line with current search. But, if the intuition about the behavior here is that it should be lower, LMK.

---

Ran into the 5K limit for pagination requests, and I think that makes sense to keep. To wit, you'll get this alert,
Screen Shot 2020-04-08 at 7 09 05 PM

instead of 

Screen Shot 2020-04-08 at 6 58 13 PM

when using `stable:`. There's probably a better way to phrase the alert, but the point is it's better than AH-A-RED-ALERT-WHAT-DOES-IT-MEAN.

---

This PR also makes `limitHit` significant to clients who might care about it (like our webapp, which depends on it to surface the 'show more' link/buttons). For pagination, we simply give the meaning to limitHit as "pagination has not reached the final cursor for this request'. 

Then we get this, as desired:

stable-show-more-short

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search.go [hunk 1]
```diff
@@ -98,6 +98,22 @@ func NewSearchImplementer(args *SearchArgs) (SearchImplementer, error) {
 		}
 	}
 
+	// If stable:truthy is specified, execute a pagination to return a stable ordering.
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search.go:101]:
```suggestion
	// If stable:truthy is specified, execute a paginated search in order to return a stable ordering.
```

File: internal/search/query/searchquery.go [hunk 2]
```diff
@@ -36,8 +36,9 @@ const (
 
 	// Temporary experimental fields:
 	FieldIndex     = "index"
-	FieldCount     = "count" // Searches that specify `count:` will fetch at least that number of results, or the full result set
-	FieldMax       = "max"   // Deprecated alias for count
+	FieldCount     = "count"  // Searches that specify `count:` will fetch at least that number of results, or the full result set
+	FieldStable    = "stable" // Forces search to return a stable result ordering for file content matches.
```
Comment (slimsag) [inline internal/search/query/searchquery.go:40]:
```suggestion
	FieldStable    = "stable" // Forces search to return a stable result ordering (currently limited to file content matches.)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -34,6 +34,7 @@ All notable changes to Sourcegraph are documented in this file.
   - Support for configuring Lightstep as a distributed tracer is deprecated and will be removed in a subsequent release. Instances that use Lightstep with Sourcegraph are encouraged to migrate to Jaeger (directions for running Jaeger alongside Sourcegraph are included in the installation instructions).
 
 - [`sourcegraph/git-extras`](https://sourcegraph.com/extensions/sourcegraph/git-extras) is now enabled by default on new instances [#3501](https://github.com/sourcegraph/sourcegraph/issues/3501)
+- search: Adding `stable:true` to a query ensures a deterministic search result order. This is an experimental parameter. It applies only to file contents, and handles requests for up to 5000 results.
```
Comment (slimsag) [inline CHANGELOG.md:37]:
```suggestion
- search: Adding `stable:true` to a query ensures a deterministic search result order. This is an experimental parameter. It applies only to file contents, and is limited to at max 5,000 results (consider using [the paginated search API](https://docs.sourcegraph.com/api/graphql/search#sourcegraph-3-9-experimental-paginated-search) if you need more than that.).
```

File: cmd/frontend/graphqlbackend/search.go [hunk 4]
```diff
@@ -130,6 +130,55 @@ func (r *schemaResolver) Search(args *SearchArgs) (SearchImplementer, error) {
...
+		}
+		if *args.First < 0 || *args.First > maxSearchResultsPerPaginationRequest {
+			return nil, fmt.Errorf("search: requested pagination 'first' value outside allowed range (0 - %d)", maxSearchResultsPerPaginationRequest)
+		}
+		pagination = &searchPaginationInfo{
+			cursor: cursor,
+			limit:  *args.First,
+		}
+	} else if args.After != nil {
+		return nil, errors.New("Search: paginated requests providing a 'after' but no 'first' is forbidden")
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search.go:177]:
```suggestion
		return nil, errors.New("search: paginated requests providing an 'after' cursor but no 'first' value is forbidden")
```

FYI: lowercase errors otherwise they do not get composed well ("Error: something bad happened: Search: paginated requests providing ...")

File: cmd/frontend/graphqlbackend/search.go [hunk 5]
```diff
@@ -130,6 +130,55 @@ func (r *schemaResolver) Search(args *SearchArgs) (SearchImplementer, error) {
...
 }
 
+// queryForStableResults transforms a query that returns a stable result
+// ordering. The transformed query uses pagination underneath the hood.
+func queryForStableResults(args *SearchArgs, queryInfo query.QueryInfo) (*SearchArgs, query.QueryInfo, error) {
+	if queryInfo.BoolValue(query.FieldStable) {
+		var stableResultCount int32
+		if _, countPresent := queryInfo.Fields()["count"]; countPresent {
+			count, _ := queryInfo.StringValue(query.FieldCount)
+			count64, err := strconv.ParseInt(count, 10, 32)
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search.go:140]: A bit weird to call this `count64` when the bit size is 32 😄I suggest renaming `count -> countStr` and `count64 -> count`

File: cmd/frontend/graphqlbackend/search.go [hunk 6]
```diff
@@ -130,6 +130,55 @@ func (r *schemaResolver) Search(args *SearchArgs) (SearchImplementer, error) {
...
+			if err != nil {
+				return nil, nil, err
+			}
+			stableResultCount = int32(count64)
+			if stableResultCount > maxSearchResultsPerPaginationRequest {
+				return nil, nil, fmt.Errorf("At most count:%d results can be requested when 'stable:' is specified. Consider removing 'stable:' or narrowing search with 'repo:'.", maxSearchResultsPerPaginationRequest)
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search.go:146]:
```suggestion
				return nil, nil, fmt.Errorf("Stable searches are limited to at max count:%d results. Consider removing 'stable:', narrowing the search with 'repo:'., or using the paginated search API.", maxSearchResultsPerPaginationRequest)
```

File: cmd/frontend/graphqlbackend/search.go [hunk 7]
```diff
@@ -130,6 +130,55 @@ func (r *schemaResolver) Search(args *SearchArgs) (SearchImplementer, error) {
...
+		} else {
+			stableResultCount = defaultMaxSearchResults
+		}
+		args.First = &stableResultCount
+		fileValue := "file"
+		// Pagination only works for file content searches, and will
+		// raise an error otherwise. If stable is explicitly set, this
+		// is implied. So, force this query to only return file content
+		// results.
+		queryInfo.Fields()["type"] = []*querytypes.Value{{String: &fileValue}}
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search.go:157]: Will the user appropriately get an error if they try to run a different type of search like `type:file type:symbol`? Or will they not get symbol results and end up surprised?

Comment (slimsag) [reply cmd/frontend/graphqlbackend/search.go:157]: (definitely makes sense to use `type:file` by default when `stable:` is specified, though)

File: cmd/frontend/graphqlbackend/search.go [hunk 8]
```diff
@@ -208,6 +257,7 @@ func (r *searchResolver) countIsSet() bool {
 }
 
 const defaultMaxSearchResults = 30
+const maxSearchResultsPerPaginationRequest = 5000
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search.go:260]:
```suggestion
const maxSearchResultsPerPaginatedRequest = 5000
```

nit: worded as-is this wording confuses me a bit; In my mind "paginated request" == "request that has been or is paginated", while "pagination request" == "a request that something become paginated"

File: cmd/frontend/graphqlbackend/search_pagination.go [hunk 9]
```diff
@@ -202,6 +202,13 @@ func (r *searchResolver) paginatedResults(ctx context.Context) (result *SearchRe
 		"Finished", cursor.Finished,
 	)
 
+	if !cursor.Finished {
+		// Clients may use limitHit to indicate partial results. For
+		// paginated requests, limitHit = true implies there is a next
+		// cursor, and more results may exist.
+		common.limitHit = true
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search_pagination.go:209]: I believe we should only do this when `stable:` is specified.

Reasoning: Paginated requests don't have a "limit" so when speaking purely about paginated requests the naming of this is a bit of a misnomer. Also, API users will have a better experience if they discover `pageInfo` first and don't end up relying on this because it matches other idiomatic GraphQL semantics around pagination.

File: doc/user/search/queries.md [hunk 10]
```diff
@@ -85,6 +85,7 @@ The following keywords can be used on all searches (using [RE2 syntax](https://g
...
 | **patterntype:literal, patterntype:regexp, patterntype:structural**  | Configure your query to be interpreted literally, as a regular expression, or a [structural search pattern](structural.md). Note: this keyword is available as an accessibility option in addition to the visual toggles. | [`test. patternType:literal`](https://sourcegraph.com/search?q=test.+patternType:literal)<br/>[`(open\|close)file patternType:regexp`](https://sourcegraph.com/search?q=%28open%7Cclose%29file&patternType=regexp) |
 | **visibility:any, visibility:public, visibility:private** | Filter results to only public or private repositories. The default is to include both private and public repositories. | [`type:repo visibility:public`](https://sourcegraph.com/search?q=type:repo+visibility:public) |
+| **stable:yes** | Ensures a deterministic result order. Applies only to file contents. Compatible with `count`, up to a maximum of 5000 results. | [`func stable:yes count:10`](https://sourcegraph.com/search?q=func+stable:yes+count:30&patternType=literal) |
```
Comment (slimsag) [inline doc/user/search/queries.md:88]:
```suggestion
| **stable:yes** | Ensures a deterministic result order. Applies only to file contents. Compatible with `count`, up to a maximum of 5000 results. Not needed when issuing paginated API queries, which are always deterministic. | [`func stable:yes count:10`](https://sourcegraph.com/search?q=func+stable:yes+count:30&patternType=literal) |
```

File: internal/search/query/types/query.go [hunk 11]
```diff
@@ -54,7 +54,10 @@ type Value struct {
 
 // Not returns whether the value is negated in the query (e.g., -value or -field:value).
 func (v *Value) Not() bool {
-	return v.syntax.Not
+	if v.syntax != nil {
+		return v.syntax.Not
+	}
+	return false
```
Comment (slimsag) [inline internal/search/query/types/query.go:60]: Q: why this change and is this covered by a test?

File: cmd/frontend/graphqlbackend/search.go (hunk 4 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:177]: Sure thing. I just move this around, didn't add it. Will update.

File: cmd/frontend/graphqlbackend/search.go (hunk 5 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:140]: `ParseInt` returns an int64 and I was "inspired" by the Go docs that just called this `i64` https://golang.org/pkg/strconv/#hdr-Numeric_Conversions. But yah, `countStr` makes sense

File: cmd/frontend/graphqlbackend/search.go (hunk 6 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:146]: Yeah I wasn't sure either. Just wanted something other than the red error. This seems fine?

File: cmd/frontend/graphqlbackend/search.go (hunk 7 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:157]: They will not get symbol results and end up "surprised" if they aren't aware of the documentation/implied nature of `stable:yes`. I can handle this better by throwing an alert if `type` is explicitly specified for anything other than `file`.

File: cmd/frontend/graphqlbackend/search.go (hunk 8 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:260]: Makes sense, I wasn't sure of this either. I initially called this '..PerPage..` but that might make unfamiliar folks think this refers to the actual pages being rendered in the webapp, whereas it's specific to the pagination functionality.

File: cmd/frontend/graphqlbackend/search_pagination.go (hunk 9 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search_pagination.go:209]: Yeah I just realized I can actually just change the limitHit part outside this function in `search_results`, based on whether `stable` exists. Wouldn't want to do that here because then pagination relies on query logic.

File: doc/user/search/queries.md (hunk 10 above)
Comment (rvantonder) [reply doc/user/search/queries.md:88]: As an outside user unfamiliar with pagination, this might scan weird like "why is this description talking about paginated API queries?". Maybe something like "Note this field has no effect If you're using the pagination API, which already ensures deterministic results."?

File: internal/search/query/types/query.go (hunk 11 above)
Comment (rvantonder) [reply internal/search/query/types/query.go:60]: [`determineResultTypes`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@97deadaaf4488b56e1b1f1500cbefbf5a086a778/-/blob/cmd/frontend/graphqlbackend/search_results.go#L1103) will check whether `type` is specified, and then retrieve that value. It will do that by calling [`StringValues`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@97deadaaf4488b56e1b1f1500cbefbf5a086a778/-/blob/internal/search/query/searchquery.go#L274). `StringValues` insists on checking whether an [expression](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@97deadaaf4488b56e1b1f1500cbefbf5a086a778/-/blob/internal/search/query/types/query.go#L48) associated with this value is `Not`. In this PR, I manually set the value of field `type` to `file`, but there is no underlying expression (`type` can't be negated anyway), but the above processing still happens. So you get a nil deref. I could set an underlying expression as well, but that means it's still possible to trigger this nil deref in the way I just described, so I'm adding this check instead.

Ideally things like `StringValues` shouldn't be rely on disjoint data types (i.e., a `*Value` and an underlying expression, which duplicates the meaning/intent of 'value') to carry out what it wants to do. Whatever information the expression contains should be lifted into the `*Value` type. Short of figuring that out and refactoring, this is an easy fix without investing more effort, since I'm rewriting the query code that'll get rid of things like this anyway.

**Comment by rvantonder** [comment]
@keegancsmith the intuition is actually directly related to `and/or` queries. Basically, because we do `evaluateLeaf` for each expression, it increases the chances of nondeterminism for each eval/search operation. When these results get merged, partial/nondet results along a single branch would create different orderings. The idea is that by adding `stable:`, we can also ensure a stable result ordering for `and/or` queries which is more prone to nondeterminism.  

I haven't done extensive testing (and this is where I will emphasize intuition and anecdotal observation), but I quickly ran into this with simple `and/or` queries where results can be jumbled up (say, one in five refreshes for a set of ~5 ish repos is a different order). By default, we'll assume this behavior is OK, but I want the `stable` option to be available in the event a user has a problem with that experience.

File: cmd/frontend/graphqlbackend/search.go (hunk 7 above)
Comment (slimsag) [reply cmd/frontend/graphqlbackend/search.go:157]: >  I can handle this better by throwing an alert if type is explicitly specified for anything other than file.

That sounds ideal / expected to me

File: doc/user/search/queries.md (hunk 10 above)
Comment (slimsag) [reply doc/user/search/queries.md:88]: That sounds good to me 👍

File: internal/search/query/types/query.go (hunk 11 above)
Comment (slimsag) [reply internal/search/query/types/query.go:60]: Makes sense 👍

**Comment by keegancsmith** [comment]
> I haven't done extensive testing (and this is where I will emphasize
> intuition and anecdotal observation), but I quickly ran into this with
> simple `and/or` queries where results can be jumbled up (say, one in
> five refreshes for a set of ~5 ish repos is a different order). By
> default, we'll assume this behavior is OK, but I want the `stable`
> option to be available in the event a user has a problem with that
> experience.

Ok that is interesting to hear. I suppose my only concern around
defaults and stable ordering are implications on performance. As long as
we don't compromise on that I'm happy :)

File: cmd/frontend/graphqlbackend/search.go (hunk 7 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go:157]: Tracking in #9715. Will do in follow diff up by release cut.

**Comment by rvantonder** [comment]
I realized that a query with `stable` will actually take precedence over an API use of pagination, meaning  the truth in the doc should be

> Note this field should be removed if you're using the pagination API, which already ensures deterministic results.

Instead of

> Note this field has no effect if you're using the pagination API, which already ensures deterministic results.

We really want the behavior of the latter (no effect), but it's tricky to just do and inject into this PR, and the fact that I missed this initially means it needs it's own test. I'll do so in a follow up PR.

**Comment by rvantonder** [comment]
Whelp, I merged this into a stacked branch, and forgot to set the base branch to master. No way to recover from this, please stamp https://github.com/sourcegraph/sourcegraph/pull/9718
